framework/contracts/account/src/migrate.rs

```diff
@@ -1,19 +1,121 @@
+use abstract_sdk::feature_objects::RegistryContract;
+use abstract_sdk::std::{account::state::ACCOUNT_ID, ACCOUNT};
+use abstract_std::account::ModuleInstallConfig;
+use abstract_std::objects::module::ModuleInfo;
+use abstract_std::objects::module_version::assert_contract_upgrade;
+use abstract_std::IBC_CLIENT;
+use abstract_std::{account::MigrateMsg, objects::AccountId};
 use abstract_std::{
-    account::MigrateMsg, objects::module_version::assert_contract_upgrade, ACCOUNT,
+    account::{
+        state::{AccountInfo, WhitelistedModules, INFO, SUSPENSION_STATUS, WHITELISTED_MODULES},
+        UpdateSubAccountAction,
+    },
+    objects::{
+        gov_type::GovernanceDetails,
+        ownership::{self},
+    },
+    registry::state::LOCAL_ACCOUNT_SEQUENCE,
 };
-use cosmwasm_std::{DepsMut, Env};
-use cw2::set_contract_version;
+use cosmwasm_std::{wasm_execute, DepsMut, Env, StdError};
+use cw2::{get_contract_version, set_contract_version};
 use semver::Version;
 
+use crate::error::AccountError;
+use crate::{
+    modules::{_install_modules, MIGRATE_CONTEXT},
+    msg::ExecuteMsg,
+};
+
 use crate::contract::{AccountResponse, AccountResult, CONTRACT_VERSION};
 
 #[cfg_attr(feature = "export", cosmwasm_std::entry_point)]
-pub fn migrate(deps: DepsMut, _env: Env, _msg: MigrateMsg) -> AccountResult {
+pub fn migrate(mut deps: DepsMut, env: Env, _msg: MigrateMsg) -> AccountResult {
     let version: Version = CONTRACT_VERSION.parse().unwrap();
 
-    assert_contract_upgrade(deps.storage, ACCOUNT, version)?;
+    // If we already have an abstract account, we just migrate like normal
+    if assert_contract_upgrade(deps.storage, ACCOUNT, version).is_ok() {
+        set_contract_version(deps.storage, ACCOUNT, CONTRACT_VERSION)?;
+        return Ok(AccountResponse::action("migrate"));
+    }
+    // else this means that we are migrating from another contract, we assert it's `account` and create an abstract account
+    let current_contract_version = get_contract_version(deps.storage)?;
+    if current_contract_version.contract != "account" {
+        return Err(AccountError::Std(StdError::generic_err("Can't migrate something other than `abstract_account`  or `account` to `abstract_account`")));
+    }
+
+    // Use CW2 to set the contract version, this is needed for migrations
     set_contract_version(deps.storage, ACCOUNT, CONTRACT_VERSION)?;
-    Ok(AccountResponse::action("migrate"))
+
+    let registry = RegistryContract::new(deps.api, &env)?;
```

Comment:
Would it be possible to share a function for this and instantiate?

Reply:
not really, I removed a lot of code that was un-necessary and all that. I might want to call instantiate directly though maybe ?

Reply:
No that's not possible as the xion authenticator is in the way. Copying the code is the best solution here I think